### PR TITLE
🎉  Add create & delete workspaces to API

### LINKED ISSUE #1735: Enable workspace creation/deletion
## Tell us about the problem you're trying to solve
Some users want to use namespaces to segment configurations in a multi-tenant setup.

## Describe the solution you’d like
Expose the feature. Right now we only rely on a default workspace

## Describe the alternative you’ve considered or used
Some manual segmentation with one namespace

### PR DESCRIPTION
**Pull request by gordalina**
## What
Closes #1735
Adds `/workspaces/create` and `/workspaces/delete`.

## How
This adds `tombstone` to the workspace.
This also adds a check to prevent returning destinations and sources that have been tombstones.

## Questions
I see that you are using `TrackingClientSingleton.get().identify();` when updating a workspace. I don't think we need to do that for the new workspaces - but I leave that decision to you.

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. WorkspacesHandler

### CODE REVIEW

**Comment by ChristopheDuong** [comment]
Hi!

Thanks for your contribution!

For your question on migration, you can take a look at this class here as an example: `airbyte-migration/src/main/java/io/airbyte/migrate/migrations/MigrationV0_14_3.java`

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:136]: Same here. We can add a boolean to `getStandardWorkspace` to indicate if you want to include the tombstone ones or not. This way you can let the config repo throw the `ConfigNotFoundException`.

WDYT?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:151]: Same idea for the prototype.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java
Comment (michel-tricot) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:136]: The good thing about adding that boolean is that people will have to make the conscious choice of including tombstoned or not. Less risk of forgetting to filter out the tombstoned

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java
Comment (michel-tricot) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:151]: Instead actually. What about adding a configRepository::getStandardWorkspaceBySlug (with the tombstone boolean as well)

File: airbyte-server/src/test/java/io/airbyte/server/handlers/WorkspacesHandlerTest.java
Comment (michel-tricot) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/WorkspacesHandlerTest.java:113]: The test is a bit brittle here. What about: Injecting the uuid generator in a protected constructor and mock it so you can control the value of the uuid and you can just do a `assertEqual(expectedWorkspaceRead, created)`

Here is an example of how we do it: `io/airbyte/server/handlers/DestinationDefinitionsHandler.java:53`

**Review by michel-tricot — COMMENTED** [review]
The change looks great! added a few comments!

**Comment by gordalina** [comment]
@michel-tricot addressed all your comments & added a 0.17.0 migration

**Comment by sherifnada** [comment]
@cgardens is it too late to include this in 0.17.0 release?

**Review by michel-tricot — APPROVED** [review]

**Comment by michel-tricot** [comment]
LGTM! You will just need to rename the migrations (we released 0.17.0 in the meantime).

**Comment by michel-tricot** [comment]
I'll merge it after that

File: airbyte-api/src/main/openapi/config.yaml
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:1212]: i think instead of exposing these in the API, you can hide them and just set them to `false` in the handler. i thin it's reasonable that these 2 fields will always be false when the workspace is created.

File: airbyte-analytics/src/main/java/io/airbyte/analytics/TrackingClientSingleton.java
Comment (cgardens) [inline airbyte-analytics/src/main/java/io/airbyte/analytics/TrackingClientSingleton.java:73]: to give you a little context, the concept of a "workspace" and a "deployment" are conflated in our system right now. that has been okay so far because there is only one workspace. the reason i mention it, is that we still to be able to use the default workspace for deployment info (even if the default workspace has been deleted).

all that being said, i think that just means we want `true` here instead of false. if the default workspace is ever deleted we still want this method to be able to access it.

File: airbyte-config/models/src/main/resources/types/StandardWorkspace.yaml
Comment (cgardens) [inline airbyte-config/models/src/main/resources/types/StandardWorkspace.yaml:12]: in the [code](https://github.com/airbytehq/airbyte/pull/2325/files#diff-f6c08873b6a8e35518a1dcc316cbc95975632c8ff27908ecffc425472b95b80bR77) i think you already check that if tombstone is not set that we treat it as false. i think that means that this field then doesn't need to be required. making it not required _should_ mean that we should be able to not need to do a migration. 

Up to you, if you feel like it should be required, then we can keep the migration and just drop the check you're doing in the config repository. LMK if that's not clear.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ArchiveHandler.java
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ArchiveHandler.java:160]: same here. i think we want true. we want the default workspace to be returned even if deleted here.

File: airbyte-server/src/main/java/io/airbyte/server/converters/ConfigFileArchiver.java
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/converters/ConfigFileArchiver.java:78]:
```suggestion
    writeConfigsToArchive(storageRoot, ConfigSchema.STANDARD_WORKSPACE, configRepository.listStandardWorkspaces(true));
```

I think we can get rid of lines 75-79 and just replace it with this. we just want to do the list of all workspaces (regardless of whether they've been deleted).

File: airbyte-server/src/main/java/io/airbyte/server/handlers/HealthCheckHandler.java
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/HealthCheckHandler.java:47]: true here too. we don't want the healthcheck to fail because the default workspace was deleted.

**Review by cgardens — CHANGES_REQUESTED** [review]
Hi Sam,

First, sorry for the slow follow up on this. There was some confusion about how was going to follow up on your PR. This was my bad.

Second, thanks for contributing this! It looks great. I have have one last round of comments. I'm hoping they are all clear and straight forward. Of course, let me know if you have any questions. 

After that, I will merge this PR (for real this time--I know we said this once before already 😉 ). I don't know what your time looks like right now, so if you can't get to these, it not the end of the world. If you just say, "sorry I'm busy right now", I will just merge the PR as is and tweak the things I mentioned in my review. You've done the hard part, there's just a couple tweaks that I'm seeing based on some knowledge of some weird assumptions that we've made.

Let me know! Looking forward to getting this merged.

**Review by cgardens — APPROVED** [review]
amazing! thanks for the fast turn around!

**Pull request merged**
